ignore/src/hgignore.rs

```diff
@@ -0,0 +1,513 @@
+            return Ok(self);
+        }
+        line = line.trim_right();
+        if line.is_empty() {
+            return Ok(self);
+        }
+        // TODO: extended patterns support, see hg help patterns
+        lazy_static! {
+            static ref EXT_PATTERN: Regex = Regex::new(r"^(\w+):").unwrap();
+        };
+        match EXT_PATTERN.captures(line.as_bytes()) {
+            Some(caps) => {
+                match &caps[1] {
+                    b"path" => debug!("HG pattern \"path:\" is not supported."),
+                    b"rootfilesin" => debug!("HG pattern \"rootfilesin:\" is not supported."),
+                    b"glob" => debug!("HG pattern \"glob:\" is not supported."),
+                    b"re" => debug!("HG pattern \"re:\" is not supported."),
+                    b"listfile" => debug!("HG pattern \"listfile:\" is not supported."),
+                    b"listfile0" => debug!("HG pattern \"listfile0:\" is not supported."),
+                    b"include" => debug!("HG pattern \"include:\" is not supported."),
+                    b"subinclude" => debug!("HG pattern \"subinclude:\" is not supported."),
+                    _ => {
+                        return match current_syntax {
+                            HgignoreSyntax::Glob => self.add_glob_line(from, line),
+                            HgignoreSyntax::Regexp => self.add_regex_line(from, line),
+                        }
+                    },
+                }
+                return Ok(self);
+            },
+            None => {
+                match current_syntax {
+                    HgignoreSyntax::Glob => self.add_glob_line(from, line),
+                    HgignoreSyntax::Regexp => self.add_regex_line(from, line),
+                }
+            },
+        }
+    }
+
+    /// doc
+    pub fn add_regex_line(
+        &mut self,
+        from: Option<PathBuf>,
+        line: &str,
+    ) -> Result<&mut HgignoreBuilder, Error> {
+        // FIXME: validate regex otherwise RegexSet build will
+        // fail. So we have to create regexes twice:
+        // during validation and during RegexSet build.
+        // Another option is to use list of regexes instead of RegexSet
+        self.regex_lines.push(Pattern::Regex {
+            from: from,
+            original: line.to_string(),
+        });
+        Ok(self)
+    }
+
+    /// doc
+    pub fn add_glob_line(
+        &mut self,
+        from: Option<PathBuf>,
+        mut line: &str,
+    ) -> Result<&mut HgignoreBuilder, Error> {
+        let mut actual;
+        let mut is_only_dir = false;
+        let original = line.to_string();
+
+        let mut literal_separator = false;
+        let has_slash = line.chars().any(|c| c == '/');
+        let mut is_absolute = false;
+        if line.starts_with("\\#") {
+            line = &line[1..];
+            is_absolute = line.chars().nth(0) == Some('/');
+        } else {
+            // FIXME: this is not the case for hgignore
```

Comment:
Can anyone describe what leading slash mean in regular glob or especially in hgignore? I can't find a case that would match glob with leading slash in hgignore, maybe just skip them?

Reply:
I think `man gitignore` says it best:

> A leading slash matches the beginning of the pathname. For example, "/*.c" matches "cat-file.c" but not "mozilla-sha1/sha1.c".

It other words, it's a form of an anchor.

Reply:
(I don't know whether it's a thing in `hgignore` or not.)

Reply:
> I don't know whether it's a thing in hgignore or not.

```
$ cat > .hgignore
 syntax: glob
 /foo
 bar

$ touch foo
$ touch bar
$ hg status
 ? .hgignore
 ? foo

$ mkdir xxx
$ touch xxx/foo
$ hg status
 ? .hgignore
 ? foo
 ? xxx/foo
```

I've tested it and haven't found a case where leading slash would match anything, I believe git uses leading slash because standard glob doesn't support leading slash at all. That's why I assumed that we can skip such rules.

Globs in .hgignore are not rooted if you want rooted glob you should use `glob:`, from the `hg help patterns`

```
To use an extended glob, start a name with "glob:". Globs are rooted at
    the current directory; a glob such as "*.c" will only match files in the
    current directory ending with ".c".
```